Warn when unique objects might be duplicated in shared libraries

DKLoehr opened the pull request:
When a hidden object is built into multiple shared libraries, each instance of the library will get its own copy. If
the object was supposed to be globally unique (e.g. a global variable or static data member), this can cause very subtle bugs.

An object might be incorrectly duplicated if it:
- Is defined in a header (so it might appear in multiple TUs), and
- Has external linkage (otherwise it's supposed to be duplicated), and
- Has hidden visibility (or else the dynamic linker will handle it)

The duplication is only a problem semantically if one of the following is true:
1. The object is mutable (the copies won't be in sync), or
2. Its initialization has side effects (it may now run more than once), or
3. The value of its address is used (different copies have different addresses).

To detect this, we add a new -Wunique-object-duplication warning. It warns on cases (1) and (2) above. To be conservative, we only warn in case (2) if we are certain the initializer has side effects, and we don't warn on `new` because the only side effect is some extra memory usage.

We don't currently warn on case (3) because doing so is prone to false positives: there are many reasons for taking the address which aren't inherently problematic (e.g. passing to a function that expects a pointer). We only run into problems if the code inspects the value of the address.

The check is currently disabled for windows, which uses its own analogue of visibility (declimport/declexport). The check is also disabled inside templates, since it can give false positives if a template is never instantiated.

### Resolving the warning
The warning can be fixed in several ways:
- If the object in question doesn't need to be mutable, it should be made const. Note that the variable must be completely immutable, e.g. we'll warn on `const int* p` because the pointer itself is mutable. To silence the warning, it should instead be `const int* const p`.
- If the object must be mutable, it (or the enclosing function, in the case of static local variables) should be made visible using `__attribute((visibility("default")))`
- If the object is supposed to be duplicated, it should be be given internal linkage.

### Testing
I've tested the warning by running it on clang itself, as well as on chromium. Compiling clang resulted in [10 warnings across 6 files](https://github.com/user-attachments/files/17908069/clang-warnings.txt), while Chromium resulted in [160 warnings across 85 files](https://github.com/user-attachments/files/17908072/chromium-warnings.txt), mostly in third-party code. Almost all warnings were due to mutable variables.

I evaluated the warnings by manual inspection. I believe all the resulting warnings are true positives, i.e. they represent potentially-problematic code where duplication might cause a problem. For the clang warnings, I also validated them by either adding `const` or visibility annotations as appropriate.

### Limitations
I am aware of four main limitations with the current warning:
1. We do not warn when the address of a duplicated object is taken, since doing so is prone to false positives. I'm hopeful that we can create a refined version in the future, however.
2. We only warn for side-effectful initialization if we are certain side effects exist. Warning on potential side effects produced a huge number of false positives; I don't expect there's much that can be done about this in modern C++ code bases, since proving a lack of side effects is difficult.
3. Windows uses a different system (declexport/import) instead of visibility. From manual testing, it seems to behave analogously to the visibility system for the purposes of this warning, but to keep things simple the warning is disabled on windows for now.
4. We don't warn on code inside templates. This is unfortuate, since it masks many real issues, e.g. a templated variable which is implicitly instantiated the same way in multiple TUs should be globally unique, but may accidentally be duplicated. Unfortunately, we found some potential false positives during testing that caused us to disable the warning for now.

DKLoehr committed:
Warn when unique objects might be duplicated in shared libraries

When a hidden object is built into multiple shared libraries, each
instance of the library will get its own copy. If
the object was supposed to be globally unique (e.g. a global
variable or static member), this can cause very subtle bugs.

An object might be incorrectly duplicated if it:
- Is defined in a header (so it might appear in multiple TUs), and
- Has external linkage (otherwise it's supposed to be duplicated), and
- Has hidden visibility (or else the dynamic linker will handle it)

The duplication is only a problem if one of the following is true:
1. The object is mutable (the copies won't be in sync), or
2. Its initialization has side effects (it may now run more than once), or
3. The value of its address is used (which one?).

To detect this, we add a new -Wunique-object-duplication warning.
It warns on cases (1) and (2) above. To be conservative, we only
warn in case (2) if we are certain the initializer has side effects,
and we don't warn on `new` because the only side effect is some
extra memory usage.

We don't currently warn on case (3) because doing so is prone to
false positives: there are many reasons for taking the address which
aren't inherently problematic (e.g. passing to a function that expects
a pointer). We only run into problems if the code inspects the value
of the address.

The check is currently disabled for windows, which uses its own analogue
of visibility (declimport/declexport). The check is also disabled inside
templates, since it can give false positives if a template is never
instantiated.

Because the warning fires in several places, included third-party
libraries, it is currently disabled-by-default until all these
instances have been fixed. It should be enabled-by-default afterwards.

DKLoehr commented:
Because this warning keeps firing in many different places, including in third-party libraries, I've disabled it by default until I can go through and fix all the instances. I think this provides a cleaner separation of concerns, so that this PR isn't clogged up with various housekeeping changes. Once the warning is cleaned up, it should be enabled-by-default.

DKLoehr commented:
Ping

DKLoehr commented:
Ping

DKLoehr commented:
Ping

zmodem commented:
I'll try to take a look. Also, @ilya-biryukov do you have someone who might want to take a look?

zmodem commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -13386,6 +13386,62 @@ void Sema::checkNonTrivialCUnion(QualType QT, SourceLocation Loc,
 
+bool Sema::GloballyUniqueObjectMightBeAccidentallyDuplicated(
+    const VarDecl *dcl) {
+  if (!dcl || !getLangOpts().CPlusPlus)
```
Could C code have similar problems?

DKLoehr replied:
C has different rules about this kind of thing, and we already have a [very similar check](https://github.com/llvm/llvm-project/blob/edf3a55bcecc8b0441a7a5fe6bda2023f86667a3/clang/include/clang/Basic/DiagnosticSemaKinds.td#L6133) for enforcing them.

zmodem commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -13386,6 +13386,62 @@ void Sema::checkNonTrivialCUnion(QualType QT, SourceLocation Loc,
+  if (!dcl || !getLangOpts().CPlusPlus)
+    return false;
+
+  // If an object is defined in a source file, its definition can't get
```
At this point it's not clear that we're looking at a definition, so maybe move this check below the "is this a definition" check below?

zmodem commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -13386,6 +13386,62 @@ void Sema::checkNonTrivialCUnion(QualType QT, SourceLocation Loc,
+  const ValueDecl *target = dcl;
+  // VarDecls and FunctionDecls have different functions for checking
+  // inline-ness, so we have to do it manually.
+  bool target_is_inline = dcl->isInline();
```
nit: LLVM style prefers camelCase, even starting with a capital letter for variables (this affects `target` and `dcl` above too)

zmodem reviewed (commented):
Some scattered thoughts:

- What about code that's not going into a shared library. Should we disable the warning then, or is the idea that the user does not enable the warning flag in that case?

- The new warning is DefaultIgnore. Do you figure it's too noisy to be on by default, or to be in some category like -Wextra? If it's just DefaultIgnore, users are unlikely to find it.

> The check is also disabled inside templates, since it can give false positives if a template is never instantiated.

Would warning during instantiation work, or is that also too noisy?

DKLoehr committed:
Address first feedback

DKLoehr commented:
If a review was requested from you and this code isn't related, apologies: For some reason attempting to rebase onto the current state of the repo ended up including a bunch of random extra commits that aren't related, and that resulted in a bunch of people automatically being added as reviewers that I can't remove. Sorry for the inconvenience.

DKLoehr commented:
@zmodem Thanks for the review! I've addressed the code comments. To your questions:
> What about code that's not going into a shared library?

My impression is that you wouldn't mark things as having hidden visibility if they're not going to a shared library at least some of the time. That said, if there's a way to check if something is being compiled into a shared library, it would make sense to add it as a condition for the warning.

> The new warning is DefaultIgnore.

I intend to make the warning on-by-default. However, it currently triggers at several places in the LLVM project, breaking the CI. Fixing them is nontrivial and cluttered up the PR, so I made it DefaultIgnore in order to keep things focused. Should the PR be accepted I plan to go through and clean things up, then default-enable the warning.

> Would warning during instantiation work?

That's the most likely solution! I need to look into it, but haven't had time yet. Since the warning is useful even without templates, I didn't think doing so should block making the PR.

DKLoehr committed:
Add documentation and a release note

DKLoehr commented:
I've now added documentation for this warning explaining what it means and how to resolve it. I also added a release note mentioning its existence.

rnk commented:
+1 to separating implementation from enablement. Enabling the warning by default will predictably generate pushback and folks may ask to revert, and it's a good best practice to be able to easily do that while leaving the implementation in place.

DKLoehr commented:
> Can you share a few of the examples where it triggers, to get a feel for what this looks like on the LLVM code?

There's a text file linked in the initial description ([here](https://github.com/user-attachments/files/18563888/clang-warnings.txt), for convenience) that summarizes the warnings I examined in clang itself. I also found a few more in other parts of the LLVM project before giving up on whack-a-mole and just turning the warning off:

```
| /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-78z5r-1/llvm-project/github-pull-requests/build-runtimes/libcxxabi/test-suite-install/include/c++/v1/__algorithm/shuffle.h:59:17: error: '__x' is mutable, has hidden visibility, and external linkage; it may be duplicated when built into a shared library [-Werror,-Wunique-object-duplication]
# |    59 |     static char __x;
```

```
/var/lib/buildkite-agent/builds/linux-56-59b8f5d88-5wwkv-1/llvm-project/github-pull-requests/third-party/benchmark/src/log.h:59:14: error: 'log_level' is mutable, has hidden visibility, and external linkage; it may be duplicated when built into a shared library [-Werror,-Wunique-object-duplication]
   59 |   static int log_level = 0;
      |              ^
/var/lib/buildkite-agent/builds/linux-56-59b8f5d88-5wwkv-1/llvm-project/github-pull-requests/third-party/benchmark/src/log.h:64:18: error: 'null_log' is mutable, has hidden visibility, and external linkage; it may be duplicated when built into a shared library [-Werror,-Wunique-object-duplication]
   64 |   static LogType null_log(static_cast<std::ostream*>(nullptr));
      |                  ^
/var/lib/buildkite-agent/builds/linux-56-59b8f5d88-5wwkv-1/llvm-project/github-pull-requests/third-party/benchmark/src/log.h:69:18: error: 'error_log' is mutable, has hidden visibility, and external linkage; it may be duplicated when built into a shared library [-Werror,-Wunique-object-duplication]
   69 |   static LogType error_log(&std::clog);
      |                  ^
```

DKLoehr commented on clang/test/SemaCXX/unique_object_duplication.h:
```diff
@@ -0,0 +1,187 @@
+namespace GlobalTest {
+  // Mutable
+  inline float disallowedGlobal1 = 3.14; // hidden-warning {{'disallowedGlobal1' is mutable, has hidden visibility, and external linkage; it may be duplicated when built into a shared library}}
+  // Same as above, but explicitly marked inline
```
You're right, it was a holdover from an earlier version of the tests. Removed.

DKLoehr commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -13374,6 +13374,62 @@ void Sema::checkNonTrivialCUnion(QualType QT, SourceLocation Loc,
 
+bool Sema::GloballyUniqueObjectMightBeAccidentallyDuplicated(
+    const VarDecl *Dcl) {
+  if (!Dcl || !getLangOpts().CPlusPlus)
```
It shouldn't be possible, but I figured it was better to check it explicitly to be safe.

zmodem replied:
If the null check is not needed, I think it should be dropped. If you look at the surrounding methods, you'll see that they don't have such checks.

DKLoehr replied:
Ok, changed.

DKLoehr commented on clang/test/SemaCXX/unique_object_duplication.h:
```diff
@@ -0,0 +1,187 @@
+/**
+ * When building shared libraries, hidden objects which are defined in header
+ * files will be duplicated, with one copy in each shared library. If the object
+ * was meant to be globally unique (one copy per program), this can cause very
+ * subtle bugs. This file contains tests for the -Wunique-object-duplication
+ * warning, which is meant to detect this.
+ * 
+ * Roughly, an object might be incorrectly duplicated if:
+ * - Is defined in a header (so it might appear in multiple TUs), and
+ * - Has external linkage (otherwise it's supposed to be duplicated), and
+ * - Has hidden visibility (or else the dynamic linker will handle it)
+ * 
+ * Duplication becomes an issue only if one of the following is true:
+ * - The object is mutable (the copies won't be in sync), or
+ * - Its initialization may has side effects (it may now run more than once), or
+ * - The value of its address is used.
+ * 
+ * Currently, we only detect the first two, and only warn on effectful
+ * initialization if we're certain there are side effects. Warning if the
+ * address is taken is prone to false positives, so we don't warn for now.
+ * 
+ * The check is also disabled on Windows for now, since it uses 
+ * dllimport/dllexport instead of visibility.
+ */
```
Fair enough; I always want more explanation when I'm looking at things, but now that it's explicitly documented we can just point to that.

DKLoehr commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -6153,6 +6153,15 @@ def warn_static_local_in_extern_inline : Warning<
   "use 'static' to give inline function %0 internal linkage">;
 
+def warn_possible_object_duplication_mutable : Warning<
+  "%0 is mutable, has hidden visibility, and external linkage; it may be "
```
I like this more.

zmodem replied:
I do think leading with the main issue, and then elaborating would be better. That may allow the reader to understand the problem quicker, is clearer if scrolling is required to read the tail of the message, etc.

I also think there's precedent. Here are two fairly random examples:

```
def warn_builtin_chk_overflow : Warning<
  "'%0' will always overflow; destination buffer has size %1,"
  " but size argument is %2">,
  InGroup<DiagGroup<"builtin-memcpy-chk-size">>;

def err_for_range_invalid: Error<
  "invalid range expression of type %0; no viable '%select{begin|end}1' "
  "function available">;
```

DKLoehr replied:
Ah, sorry, I realized my comment was confusing. I meant to say that I like your version more, so I changed it. Apologies!

DKLoehr commented on clang/test/SemaCXX/unique_object_duplication.h:
```diff
@@ -0,0 +1,187 @@
+  static double allowedStatic2 = init_dynamic(2);
+  static char allowedStatic3 = []() { return allowedStatic1++; }();
+
+  static constexpr int allowedStatic4 = 0;
+  static const float allowedStatic5 = 1;
+  static constexpr int allowedStatic6 = init_constexpr(2);
+  static const int allowedStatic7 = init_constexpr(3);
```
In this case the lines are slightly different. It's probably redundant but I don't think it hurts to check the different variations (const/constexpr, literal/constexpr initialization).

DKLoehr commented on clang/test/SemaCXX/unique_object_duplication.h:
```diff
@@ -0,0 +1,187 @@
+  static double allowedStatic2 = init_dynamic(2);
+  static char allowedStatic3 = []() { return allowedStatic1++; }();
+
+  static constexpr int allowedStatic4 = 0;
+  static const float allowedStatic5 = 1;
+  static constexpr int allowedStatic6 = init_constexpr(2);
+  static const int allowedStatic7 = init_constexpr(3);
```
Makes sense. I'll keep one around just to make sure it doesn't warn, but the rest are probably redundant.

DKLoehr committed:
Address feedback again

DKLoehr commented:
Addressed feedback, responded where appropriate. I've also done some digging into how the warning should work on Windows: it seems like it'll be very straightforward (just replace "has hidden visibility" with "doesn't have a dll import/export attribute"), but I'll leave that for a future PR to keep each the PRs focused.

zmodem commented:
Sounds good. This basically seems fine to me overall. I've just added two more comments.

DKLoehr committed:
Remove redundant null check

zmodem reviewed (approved):
lgtm

Unless there are any further comments, I'll push the Merge button on Monday.

zmodem commented:
Looks like some buildbots are unhappy. I'll back it out for now.

DKLoehr commented:
Thanks @zmodem for merging (and also reverting). Follow-up PR with the fix is #125526.
